hrc_speech_prediction/models.py

```diff
@@ -90,12 +150,183 @@ def fit(self, X_context, X_speech, labels):
+                                          self.actions)
+        else:
+            self.speech_model = self.model_generator(
+                self.actions, features="speech").partial_fit(
+                    self._X_context, x_u, action, classes=self.actions)
+
+    def predict(self,
+                cntxt,
+                utter=None,
+                model="both",
+                exclude=None,
+                plot=False,
+                return_probs=False):
+
+        if model == "context" is not None:
+            probs = self.get_context_probs(cntxt)
+        elif model == "speech" and utter is not None:
+            probs = self.get_speech_probs(utter)
+        elif model == "both" and utter is not None:
+            context_probs = self.get_context_probs(cntxt)
+            speech_probs = self.get_speech_probs(utter)
+        else:
+            raise "Error, bad inputs to predict()!"
+
+        probs = np.multiply(context_probs, speech_probs)
+
+        action = self.get_probable_action(probs, exclude)
+        # self.curr = self.curr._get_next_node(action)
+
+        if plot and model == "both":
+            all_probs = np.vstack((speech_probs, context_probs, probs))
+            names = ["Speech", "Context", "Combined"]
+
+            plt.gcf().clear()
+            plots.plot_predict_proba(
+                all_probs, self.actions, utter, model_names=names)
+            plt.show(block=False)
+        if return_probs and model == "both":
+            return action, speech_probs, context_probs, probs
+        else:
+            return action, probs
+
+    def get_speech_probs(self, utter):
+        "Takes a speech utterance and returns probabilities for each \
+            possible action on the speech model alone"
+
+        if isinstance(utter, str):
+            x_u = self._vectorizer.transform([utter])
+        else:
+            x_u = utter  # Then the input is an numpy array already
+
+        probs = self.speech_model._predict_proba(self._X_context, x_u)[0]
+
+        return self._apply_eps(self._speech_eps, probs)
+
+    def get_context_probs(self, cntxt):
+        curr = self.context_model.curr(cntxt)
+        probs = curr._get_context_probs(self._context_eps, self.actions)
+
+        return self._apply_eps(self._context_eps, probs)
+
+    def _apply_eps(self, eps, p):
+        u = np.array([1.0 / len(self.actions) for i in self.actions])
+
+        return (1.0 - eps) * p + (u * eps)
+
+    def get_probable_action(self, probs, exclude):
+        # prob indices from highest to lowest
+        high_low_probs_idx = (-probs).argsort()
+
+        # Check to make sure we haven't tried to take action already
+        if exclude:
+            for i in high_low_probs_idx:
+                act = self.actions[i]
+                if act not in exclude:
+                    return act
+            raise "Error, all actions excluded!"
+        else:
+            return self.actions[high_low_probs_idx[0]]
+
+    def plot_predicitions(self,
```

Comment:
You actually had it! But it now needs to make use of the one in the plot module…